consuming: "invalid memory address or nil pointer dereference"

I am able to read messages from one broker, but not another (same Kafka version `0.10.2`). I am able to consume messages from all using Clojure/Java.

Here is what I am getting with `segmentio/kafka-go`:

```go
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13788d2]

goroutine 85 [running]:
github.com/segmentio/kafka-go.(*Conn).SetReadDeadline(0x0, 0xbe7644d72f0448ba, 0x26204df38, 0x16aef00, 0xbe7644d72f0448ba, 0x26204df38)
	/../go/src/github.com/segmentio/kafka-go/conn.go:168 +0x22
github.com/segmentio/kafka-go.(*reader).read(0xc4204e8000, 0x16793e0, 0xc420064300, 0x0, 0x0, 0x0, 0x1675460, 0x14c3c60)
	/../go/src/github.com/segmentio/kafka-go/reader.go:692 +0xcc
github.com/segmentio/kafka-go.(*reader).run(0xc4204e8000, 0x16793e0, 0xc420064300, 0x0, 0xc4204e4440)
	/../go/src/github.com/segmentio/kafka-go/reader.go:585 +0x311
created by github.com/segmentio/kafka-go.(*Reader).start
	/../go/src/github.com/segmentio/kafka-go/reader.go:476 +0x219
```

how do I go about debugging this? Again, same consuming code with `segmentio/kafka-go` does work with one broker, but not with another. The only difference in config is a hostname.

--- a/error.go
+++ b/error.go
@@ -94,7 +94,7 @@ func (e Error) Title() string {
 	case Unknown:
 		return "Unknown"
 	case OffsetOutOfRange:
-		return "Ouffset Out Of Range"
+		return "Offset Out Of Range"
 	case InvalidMessage:
 		return "Invalid Message"
 	case UnknownTopicOrPartition:

--- a/reader.go
+++ b/reader.go
@@ -550,6 +550,7 @@ func (r *reader) run(ctx context.Context, offset int64, join *sync.WaitGroup) {
 			r.withErrorLogger(func(log *log.Logger) {
 				log.Printf("error initializing the kafka reader for partition %d of %s: %s", r.partition, r.topic, OffsetOutOfRange)
 			})
+			continue
 		default:
 			// Wait 4 attempts before reporting the first errors, this helps
 			// mitigate situations where the kafka server is temporarily

--- a/reader_test.go
+++ b/reader_test.go
@@ -45,6 +45,11 @@ func TestReader(t *testing.T) {
 			scenario: "calling Stats returns accurate stats about the reader",
 			function: testReaderStats,
 		},
+
+		{ // https://github.com/segmentio/kafka-go/issues/30
+			scenario: "reading from an out-of-range offset waits until the context is cancelled",
+			function: testReaderOutOfRangeGetsCanceled,
+		},
 	}
 
 	for _, test := range tests {
@@ -221,6 +226,31 @@ func testReaderStats(t *testing.T, ctx context.Context, r *Reader) {
 		t.Error("bad stats:")
 		t.Log("expected:", expect)
 		t.Log("found:   ", stats)
+	}
+}
+
+func testReaderOutOfRangeGetsCanceled(t *testing.T, ctx context.Context, r *Reader) {
+	prepareReader(t, ctx, r, makeTestSequence(10)...)
+
+	const D = 100 * time.Millisecond
+	t0 := time.Now()
+
+	ctx, cancel := context.WithTimeout(ctx, D)
+	defer cancel()
+
+	if err := r.SetOffset(42); err != nil {
+		t.Error(err)
+	}
+
+	_, err := r.ReadMessage(ctx)
+	if err != context.DeadlineExceeded {
+		t.Error("bad error:", err)
+	}
+
+	t1 := time.Now()
+
+	if d := t1.Sub(t0); d < D {
+		t.Error("ReadMessage returned too early after", d)
 	}
 }
 

